### ISSUE TITLE
Secrets need to be stored better

### ISSUE THREAD

**Issue by TCFox** [NZFurs/Login#67]
Currently using `IdentityServer4.Validators.HashedSharedSecretValidator` to validate client secrets, which is pretty garbage (non-salted SHA256). Need to migrate to an Argon2i based solution.

**Comment by TCFox** [comment]
Need to work on design document and [`IdentityServer4.Validators.ISecretValidator`](https://github.com/IdentityServer/IdentityServer4/blob/master/src/IdentityServer4/src/Validation/ISecretValidator.cs) implementation.

Current plan is to store the secret as follows:

* `secretid` = Identifies which secret this is referring to (as clients can have multiple secrets). 32-bit random value, can be disclosed to client.
* `Salt` = Protects hash against pre-computation attacks and forces offline cracking to be performed against one hash at a time. 256-bit random value, must not be disclosed to client.
* `Secret` = The secret used to validate against the Client ID. 256-bit value, must be disclosed to client.
* `HMAC Key` = Used to sign every secret generated for all clients, so the server does not waste time fetching/validating the secret if it is clearly invalid. Stored in config on the disk, loaded into memory on load. Not disclosed to clients.

| Client ID | Secret ID | Salt | Hashed Secret |
|-|-|-|-|
| Reference to Client (`int`) | Random 32-bit (`int`) | Random 256-bit (`byte[]`) | Blake2b(`id`+`salt`+`secretbytes`) |

Secret is generated for the RP in the following format (fixed-length, concatenated):
`secretid`+`secretbytes`+`HMAC(id+secretbytes, hmac_key)`

Will pass these on to @Soatok to verify

**Comment by soatok** [comment]
I only have one suggested change:

```diff
- secretid+secretbytes+HMAC(id+secretbytes, hmac_key)
+ secretid+secretbytes+HMAC(secretid+secretbytes, hmac_key)
```

It's important (although probably needlessly pedantic) to keep your variables consistent when designing systems. Otherwise, you risk creating a bug because what you *meant* may seem obvious to you, but your intent may not be understood by another implementor, and then things go awry from there.

I assume the protocol for verifying a secret looks like this:

1. Digest incoming string into `secretid`, `secretbytes`, and `tag` (the latter being the HMAC output)
2. Recalculate `secretid+secretbytes+HMAC(secretid+secretbytes, hmac_key)`, store the result in `calc`
3. Compare `calc` with `tag` with a constant-time compare function. If it doesn't match, abort.
4. Look up the DB Table based on `secretid` (and `expiry` against the current time). If it's not found, abort.
5. Recalculate `BLAKE2b(secretid+salt+secretbytes)`, compare securely against `Hashed Secret` stored in the database. If it doesn't match, abort.
6. If we arrive here without aborting, authentication successful.

## Analysis

### Side-Channel Cryptanalysis

The HMAC will fail fast, preventing invalid inputs from ever being used in the database lookup. This means that, without the HMAC key, attackers will not be able to make "trial" guesses at `secretbytes` to find existent/nonexistent values in the database (i.e. through timing leaks).

Therefore, timing leaks downstream in this protocol are irrelevant, since they won't be exploitable unless you already possess the key.

### Algorithm Choice

Since the inner BLAKE2b has a 256-bit random salt, and `secretbytes` is a long random, high-entropy string, this is sufficient to mitigate offline attacks without requiring a KDF or password hashing algorithm (i.e. bcrypt, scrypt, Argon2id).

If low-entropy inputs were provided (i.e. passwords), a fast hash like BLAKE2b would be inappropriate.

### Overall

This design should provide secure authentication, provided the keys are properly managed.

**Comment by TCFox** [comment]
Fantastic, thanks heaps @soatok! I'll update the above and add a note regarding the management of keys in this instance. :+1:

**Comment by TCFox** [comment]
Current work is in `feature/improved-secrets-validator`

**Comment by TCFox** [comment]
Merged in b949e47.
